cirq-ft/cirq_ft/algos/arithmetic_gates.py

```diff
@@ -130,6 +131,135 @@ def _t_complexity_(self) -> infra.TComplexity:
+        x_msb, x_lsb = x
+        y_msb, y_lsb = y
+
+        def _cswap(c: cirq.Qid, a: cirq.Qid, b: cirq.Qid, q: cirq.Qid) -> cirq.OP_TREE:
+            yield cirq.CNOT(a, b)
+            yield and_gate.And(adjoint=self.adjoint).on(c, b, q)
+            yield cirq.CNOT(q, a)
+            yield cirq.CNOT(a, b)
+
+        def _decomposition():
+            yield cirq.X(ancilla[0])
+            yield cirq.CNOT(y_msb, x_msb)
+            yield cirq.CNOT(y_lsb, x_lsb)
+            yield from _cswap(x_msb, x_lsb, ancilla[0], ancilla[1])
+            yield from _cswap(x_msb, y_msb, y_lsb, ancilla[2])
+            yield cirq.CNOT(y_lsb, x_lsb)
+
+        if self.adjoint:
+            yield from reversed(tuple(cirq.flatten_to_ops(_decomposition())))
+        else:
+            yield from _decomposition()
+
+    def __repr__(self) -> str:
+        return f'cirq_ft.algos.BiQubitsMixer({self.adjoint})'
+
+    def __pow__(self, power: int) -> cirq.Gate:
+        if power == 1:
+            return self
+        if power == -1:
+            return BiQubitsMixer(adjoint=self.adjoint ^ True)
+        return NotImplemented  # coverage: ignore
+
+    def _t_complexity_(self) -> infra.TComplexity:
+        if self.adjoint:
+            return infra.TComplexity(clifford=18)
+        return infra.TComplexity(t=8, clifford=28)
+
+
+@attr.frozen
+class SingleQubitCompare(infra.GateWithRegisters):
+    """Applies U|a>|b>|0>|0> = |a> |a=b> |(a<b)> |(a>b)>
+
+    Source: (FIG. 3) in https://www.nature.com/articles/s41534-018-0071-5#Sec8
+    """
+
+    adjoint: bool = False
+
+    @cached_property
+    def registers(self) -> infra.Registers:
+        return infra.Registers.build(a=1, b=1, less_than=1, greater_than=1)
+
+    def __repr__(self) -> str:
+        return f'cirq_ft.algos.SingleQubitCompare({self.adjoint})'
+
+    def decompose_from_registers(
+        self, *, context: cirq.DecompositionContext, **quregs: Sequence[cirq.Qid]
+    ) -> cirq.OP_TREE:
+        def _decomposition() -> Iterator[cirq.Operation]:
+            (a,), (b,), (less_than,), (greater_than,) = (
+                quregs['a'],
+                quregs['b'],
+                quregs['less_than'],
+                quregs['greater_than'],
+            )
+            yield and_gate.And([0, 1], adjoint=self.adjoint).on(a, b, less_than)
+            yield cirq.CNOT(less_than, greater_than)
+            yield cirq.CNOT(b, greater_than)
+            yield cirq.CNOT(a, b)
+            yield cirq.CNOT(a, greater_than)
+            yield cirq.X(b)
+
+        if self.adjoint:
+            yield from reversed(tuple(_decomposition()))
+        else:
+            yield from _decomposition()
+
+    def __pow__(self, power: int) -> cirq.Gate:
+        if power % 2 == 0:
+            return cirq.IdentityGate(4)
+        adjoint = power < 0
+        if adjoint:
+            return SingleQubitCompare(adjoint=self.adjoint ^ True)
+        return self
```

Comment:
What if power is `2*np.pi/3`?

Reply:
it's only defined for integer powers.. raising an error otherwise